### PR TITLE
provider: Use ConfigureContextFunc

### PR DESCRIPTION
**Pull request by ggsood**
EC Provider refactoring to use new ConfigureContextFunc

## Description
The change will refactor the EC provider code to use new ConfigureContextFunc which provides better diagnosis capabilities.

## Related Issues
https://github.com/elastic/terraform-provider-ec/issues/34

## Motivation and Context
Use the updated terraform sdk functionalities.

## How Has This Been Tested?
Locally by consuming the new provider and creating resources at Elastic cloud.

## Types of Changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Refactoring (improves code quality but has no user-facing effect)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation

## Readiness Checklist

- [x] My code follows the code style of this project
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed

### CODE REVIEW

File: ec/provider_config.go [hunk 1]
```diff
@@ -33,22 +35,36 @@ const (
...
+	// Warning or errors can be collected in a slice type
+	var diags diag.Diagnostics
 	if err != nil {
-		return nil, err
+		diags = append(diags, diag.Diagnostic{
+			Severity: diag.Error,
+			Summary:  "Unable to get EC client",
+			Detail:   "Invalid timeout value for EC client",
+		})
+		return nil, diags
```
Comment (karencfv) [inline ec/provider_config.go:49]: It's a good idea to return the errors that we are getting from other functions as they are. This way, what went wrong is clearer. Let's use the `FromErr` function instead :) 
```suggestion
		return nil, diag.FromErr(err)
```

File: ec/provider_config.go [hunk 2]
```diff
@@ -33,22 +35,36 @@ const (
...
 	})
+
 	if err != nil {
-		return nil, err
+		diags = append(diags, diag.Diagnostic{
+			Severity: diag.Error,
+			Summary:  "Unable to get EC client",
+			Detail:   "Invalid credentials for EC client",
+		})
+		return nil, diags
```
Comment (karencfv) [inline ec/provider_config.go:64]:
```suggestion
		return nil, diag.FromErr(err)
```

File: ec/provider_config.go [hunk 3]
```diff
@@ -58,6 +74,18 @@ func configureAPI(d *schema.ResourceData) (interface{}, error) {
...
 	})
+
+	if err != nil {
+		diags = append(diags, diag.Diagnostic{
+			Severity: diag.Error,
+			Summary:  "Unable to create EC client",
+			Detail:   "Unable to authenticate user for EC client",
+		})
+
+		return nil, diags
```
Comment (karencfv) [inline ec/provider_config.go:85]:
```suggestion
		return nil, diag.FromErr(err)
```

File: ec/provider_config.go [hunk 4]
```diff
@@ -58,6 +74,18 @@ func configureAPI(d *schema.ResourceData) (interface{}, error) {
...
+	}
+
+	return api, diags
```
Comment (karencfv) [inline ec/provider_config.go:88]: We can return `nil` directly here as all errors are fatal and we are not collecting a series of diagnostics yet.
```suggestion
	return api, nil
```

File: ec/provider_config.go [hunk 5]
```diff
@@ -33,22 +35,36 @@ const (
 )
 
 // configureAPI implements schema.ConfigureFunc
-func configureAPI(d *schema.ResourceData) (interface{}, error) {
+func configureAPI(ctx context.Context, d *schema.ResourceData) (interface{}, diag.Diagnostics) {
 	timeout, err := time.ParseDuration(d.Get("timeout").(string))
+
+	// Warning or errors can be collected in a slice type
+	var diags diag.Diagnostics
```
Comment (karencfv) [inline ec/provider_config.go:42]: This won't be necessary any more if we use the `FromErr` function, and we are not building a collection of several diagnostics

File: ec/provider_config.go (hunk 2 above)
Comment (ggsood) [reply ec/provider_config.go]: Thanks @karencfv for the review 😄 have addressed the suggested changes.

File: ec/provider_config.go [hunk 6]
```diff
@@ -18,7 +18,9 @@
 package ec
 
 import (
+	"context"
 	"fmt"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
```
Comment (marclop) [inline ec/provider_config.go:23]: This import should be moved to line 28, and preserve the space between the stdlib imports and the external ones. It's something that we should have in our linter. I'll add that linter check so this doesn't happen again.

Comment (ggsood) [reply ec/provider_config.go]: Thanks @marclop for the review, have corrected the order of imports.

**Comment by marclop** [comment]
running `make format` should fix the imports and produce this patch:

```diff
diff --git a/ec/provider_config.go b/ec/provider_config.go
index 2ba3524..4d28885 100644
--- a/ec/provider_config.go
+++ b/ec/provider_config.go
@@ -26,8 +26,8 @@ import (
 
 	"github.com/elastic/cloud-sdk-go/pkg/api"
 	"github.com/elastic/cloud-sdk-go/pkg/auth"
-	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 )
 
 const (

```

**Comment by ggsood** [comment]
Thanks @marclop have fixed the lint errors.

**Pull request merged**